### PR TITLE
zephyr: ll: ipc4: idc: ace: Add support for a watchdog in ll domain

### PR DESCRIPTION
**Pull request by softwarecki**
Added:
-  a new LL_WATCHDOG option to the platform Kconfig. 
-  watchdog support to zephyrs ll scheduler domain.
- ipc4 notification structure for a watchdog timeout. 

Created library header for watchdog in ll. To the idc added a new notification about the crash of the slave core. It is directed to the primary core. Only failure detected by watchdog is supported at this moment. Finally, added watchdog support for the ace platform.

### CODE REVIEW

File: src/platform/intel/ace/lib/watchdog.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+DECLARE_SOF_UUID("wdt", wdt_uuid, 0x13c8bc59, 0xc4fa, 0x4ad1,
+		 0xb9, 0x3a, 0xce, 0x97, 0xcd, 0x30, 0xac, 0xc7);
+
+DECLARE_TR_CTX(wdt_tr, SOF_UUID(wdt_uuid), LOG_LEVEL_INFO);
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adsp_watchdog), okay)
+const struct device *const watchdog = DEVICE_DT_GET(DT_NODELABEL(adsp_watchdog));
+struct ipc_msg slave_timeout_ipc;
+
+void wdt_slave_core_action_on_timeout()
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:32]: we usually put `void` as argument but I guess current compilers don't mind empty argument lists either. Also, does this function have to be global?

File: src/platform/intel/ace/lib/watchdog.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+	/* Send Watchdog Timeout IPC notification */
+	ipc4_notification_watchdog_init(&notif, cpu_get_id(), true);
+	intel_adsp_ipc_send_message_emergency(INTEL_ADSP_IPC_HOST_DEV,
+					      notif.primary.dat, notif.extension.dat);
+}
+
+static void wdt_timeout(const struct device *dev, int core)
+{
+	if (!core)
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:57]: `core == PLATFORM_PRIMARY_CORE_ID`

File: src/platform/intel/ace/lib/watchdog.c [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+		wdt_master_core_action_on_timeout();
+	else
+		wdt_slave_core_action_on_timeout();
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:60]: "primary" and "secondary" - throughout the code please

File: src/platform/intel/ace/lib/watchdog.c [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	msg.header = IDC_MSG_SLAVE_CORE_CRASHED | IDC_SCC_CORE(cpu_get_id()) |
+		     IDC_SCC_REASON(IDC_SCC_REASON_WATCHDOG);
+	msg.extension = 0;
+	msg.core = 0;
+	msg.size = 0;
+	msg.payload = NULL;
+	idc_send_msg(&msg, IDC_NON_BLOCKING);
+}
+
+void wdt_master_core_action_on_timeout()
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:45]: `static`

File: src/platform/intel/ace/lib/watchdog.c [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+static void wdt_timeout(const struct device *dev, int core)
+{
+	if (!core)
+		wdt_master_core_action_on_timeout();
+	else
+		wdt_slave_core_action_on_timeout();
+}
+
+void watchdog_init(void)
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:63]: would it be better to have all functions in this file use the same naming - either `watchdog_*` or `wdt_*`?

File: src/platform/intel/ace/lib/watchdog.c [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+/* 13c8bc59-c4fa-4ad1-b93a-ce97cd30acc7 */
+DECLARE_SOF_UUID("wdt", wdt_uuid, 0x13c8bc59, 0xc4fa, 0x4ad1,
+		 0xb9, 0x3a, 0xce, 0x97, 0xcd, 0x30, 0xac, 0xc7);
+
+DECLARE_TR_CTX(wdt_tr, SOF_UUID(wdt_uuid), LOG_LEVEL_INFO);
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(adsp_watchdog), okay)
+const struct device *const watchdog = DEVICE_DT_GET(DT_NODELABEL(adsp_watchdog));
+struct ipc_msg slave_timeout_ipc;
```
Comment (lyakh) [inline src/platform/intel/ace/lib/watchdog.c:30]: both `static`?

File: src/idc/idc.c [hunk 7]
```diff
@@ -273,6 +274,26 @@ static void idc_prepare_d0ix(void)
...
+	const uint32_t reason = (header >> IDC_SCC_REASON_SHIFT) & IDC_SCC_REASON_MASK;
+
+	switch (reason) {
+#if IS_ENABLED(CONFIG_LL_WATCHDOG)
+	case IDC_SCC_REASON_WATCHDOG:
+		watchdog_slave_core_timeout(core);
+		break;
+#endif
+	default:
+		;
```
[the remark this answers is not in the document]
Comment (softwarecki) [reply src/idc/idc.c:293]: In the future, I plan to expand the switch to support IDC_SCC_REASON_EXCEPTION.

File: src/include/sof/lib/watchdog.h [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/* SPDX-License-Identifier: BSD-3-Clause
```
Comment (dbaluta) [inline src/include/sof/lib/watchdog.h:1]: Care to add some explanation in the commit message on *why* this is needed and how it will be used?

File: src/samples/audio/detect_test.c [hunk 9]
```diff
@@ -27,8 +27,10 @@
 #include <ipc/control.h>
```
Comment (dbaluta) [inline src/samples/audio/detect_test.c:27]: This are some sort of fixes. Usually they should come first in a patch series.

File: src/include/ipc4/notification.h [hunk 10]
```diff
@@ -45,7 +45,7 @@ enum sof_ipc4_notification_type {
 #define SOF_IPC4_GLB_NOTIFY_DIR_MASK		BIT(29)
 #define SOF_IPC4_REPLY_STATUS_MASK		0xFFFFFF
 #define SOF_IPC4_GLB_NOTIFY_TYPE_SHIFT		16
-#define SOF_IPC4_GLB_NOTIFY_MSG_TYPE_SHIFT		24
+#define SOF_IPC4_GLB_NOTIFY_MSG_TYPE_SHIFT	24
```
Comment (dbaluta) [inline src/include/ipc4/notification.h:49]: This is not related to this patch? Could be easily added as a separate patch even a separat PR.

File: src/platform/intel/ace/platform.c [hunk 11]
```diff
@@ -101,6 +102,10 @@ int platform_init(struct sof *sof)
 	trace_point(TRACE_BOOT_PLATFORM_IPC);
 	ipc_init(sof);
 
+#if IS_ENABLED(CONFIG_LL_WATCHDOG)
+	watchdog_init();
+#endif
```
Comment (dbaluta) [inline src/platform/intel/ace/platform.c:107]: there is no need to protect this with `CONFIG_LL_WATCHDOG` as `watchdog_init` exists but it's a noop.

File: src/include/sof/lib/watchdog.h (hunk 8 above)
Comment (softwarecki) [reply src/include/sof/lib/watchdog.h:1]: Added description in header file.

File: src/platform/Kconfig [hunk 12]
```diff
@@ -622,4 +622,11 @@ config INCOHERENT
 
 endif
 
+config LL_WATCHDOG
+	bool "Enable watchdog support in ll scheduler"
+	default n
+	help
+	  Select if you want to protect ll scheduler with a watchdog timer
+	  and it is supported by yours platform.
```
Comment (marcinszkudlinski) [inline src/platform/Kconfig:630]: and it is supported by yours platform.

sound strange, maybe a kconfig dependency for platform that support watchdog?

File: src/platform/intel/ace/lib/watchdog.c [hunk 13]
```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Comment (marcinszkudlinski) [inline src/platform/intel/ace/lib/watchdog.c:1]: comments like // are banned

File: src/platform/intel/ace/lib/watchdog.c [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+		watchdog_primary_core_action_on_timeout();
+	else
+		watchdog_secondary_core_action_on_timeout();
+}
+
+void watchdog_init(void)
+{
+	int i, ret;
+	const struct wdt_timeout_cfg watchdog_config = {
+		.window.max = 2 * LL_TIMER_PERIOD_US / 1000,
```
Comment (marcinszkudlinski) [inline src/platform/intel/ace/lib/watchdog.c:80]: Maybe config, at least a macro instead of "magic" 2 times a period?

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]

**Review by lyakh — APPROVED** [review]

**Comment by lgirdwood** [comment]
@marcinszkudlinski updates good for you now ?

File: src/platform/intel/ace/lib/watchdog.c [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+LOG_MODULE_REGISTER(wdt, CONFIG_SOF_LOG_LEVEL);
+
+/* 13c8bc59-c4fa-4ad1-b93a-ce97cd30acc7 */
+DECLARE_SOF_UUID("wdt", wdt_uuid, 0x13c8bc59, 0xc4fa, 0x4ad1,
+		 0xb9, 0x3a, 0xce, 0x97, 0xcd, 0x30, 0xac, 0xc7);
+
+DECLARE_TR_CTX(wdt_tr, SOF_UUID(wdt_uuid), LOG_LEVEL_INFO);
+
+static const struct device *const watchdog = DEVICE_DT_GET(DT_NODELABEL(adsp_watchdog));
+static struct ipc_msg slave_timeout_ipc;
```
Comment (kv2019i) [inline src/platform/intel/ace/lib/watchdog.c:31]: Please let's not add new usages of "slave". "secondary_timeout_ipc" or something like that...

**Review by kv2019i — CHANGES_REQUESTED** [review]
Looks good. One s/slave/secondary/ change needed still.

**Review by marcinszkudlinski — APPROVED** [review]

**Review by abonislawski — APPROVED** [review]
+1 but note that this is problematic for debug

**Review by kv2019i — APPROVED** [review]
Looks good now!
